kiali/kiali
# KIALI-1186 Use oauth token when contacting the k8s master

**Describe the change**

If we are using the Oauth login mechanism for OpenShift, we will now use this token when contacting the k8s master. This means the permissions the user has in the OpenShift cluster will affect what they can access in Kiali.

If a user is not using the Oauth login mechanism, then we use the service accounts token when contacting the k8s master.

This basically brings us up to the 'mulitenancy' support that we want to have with OpenShift.

**Issue reference**

https://issues.jboss.org/browse/KIALI-1186

**Documentation**

We do need documentation to be updated. Will submit a PR for docs once this PR has been reviewed.

**Known Issues**

There are a few issues I have seen so far:

1) if you are in the graph seeing one namespace and then later login with a user that doesn't have access to the namespace you are shown an error.
2) if your user doesn't have access to the kiali-system namespace you will get an error about not being able to generate the grafana links.

These are also issues that we have if you were running with reduced permissions. I would rather fix those in a separate jira and keep this one more focused on the task we are trying to do here.

User who are logging into Kiali may not have all the permissions that the `kiali` role grants, so we might experience some issues with access denied problems (and will need to work towards making them work better). The easier way to get around this is to grant the user the `kiali` role for that namespace (`oc adm policy add-role-to-user kiali ${USER} -n ${NAMESPACE}`)


**Other Considerations**
We should also see how this affects performance. Before we were reusing the k8s client across all connections and now we are creating new ones per user requests. Its unknown at this point how much of a performace hit this will cause.

**How to test**
1) build Kiali like you would if you were using oauth authentication 
ie `AUTH_STRATEGY=openshift make openshift-deploy`

2) log into openshift using a user account. You should only be displayed data that you have access to. If you want to grant a user access to another namespace, you can grant them the `kiali` role:
`oc adm policy add-role-to-user kiali ${USER} -n ${NAMESPACE}`
